### ISSUE TITLE
User Secrets not loading Microsoft.Extensions.Hosting 6.0

### ISSUE THREAD

**Issue by devhl-labs** [dotnet/runtime#63246]
### Description

The host is not loading secrets from the user secrets nor from the command line. 

### Reproduction Steps

https://github.com/devhl-labs/Hosting-6.0-Issue-Repro/blob/daabb5c610c640a76fae9d2bc0391fcc620424db/WorkerService1/WorkerService1/Program.cs#L20

See the repo above, or create a new worker service project. Observe the IConfigurationBuilder sources count be different when using Microsoft.Extensions.Hosting 5.0 and 6.0. Look at the HostBuilderContext.Configuration.Providers and observe that what is missing is user secrets and CommandLineConfigurationProvider.

### Expected behavior

The loaded secrets should be the same.

### Actual behavior

Missing secrets configurations when using 6.0.

### Regression?

Yes, this worked under 5.0

**Comment by devhl-labs** [comment]
A partial fix is to manually add this to the csproj. It will add the user secrets, but not the CommandLineConfigurationProvider.

`<PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="6.0.0" />`

**Comment by eerhardt** [comment]
> Observe the IConfigurationBuilder sources count be different when using Microsoft.Extensions.Hosting 5.0 and 6.0. Look at the HostBuilderContext.Configuration.Providers and observe that what is missing is user secrets and CommandLineConfigurationProvider.

1. The `CommandLineConfigurationProvider` provider is no longer there due to this change: #50447, it changed a `null` check to also check for the command line args to have `Length > 0`:

https://github.com/dotnet/runtime/blob/312c66f1fc2f749f56612999cb1adab9ca7fde59/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L219-L222

Since there are no command line arguments, it doesn't seem necessary to add the provider.

2. But the more important question is why isn't the user secrets there? The reason it isn't showing up is because the assembly level attribute isn't getting injected into the assembly. The place this _should_ happen is in our build targets:

https://github.com/dotnet/runtime/blob/8ce1c54005c57a0fa835fb0aa33ece45300b4a8e/src/libraries/Microsoft.Extensions.Configuration.UserSecrets/src/build/netstandard2.0/Microsoft.Extensions.Configuration.UserSecrets.targets#L8-L11

When using the `5.0.0` version of the package, this `.targets` file is getting imported. But when I use the `6.0.0` version, it doesn't get imported for some reason that I can't figure out. I am able to prove this by looking at the `project.assets.json` file:

#### 5.0.0

```
      "Microsoft.Extensions.Configuration.UserSecrets/5.0.0": {
        "type": "package",
        "dependencies": {
          "Microsoft.Extensions.Configuration.Abstractions": "5.0.0",
          "Microsoft.Extensions.Configuration.Json": "5.0.0",
          "Microsoft.Extensions.FileProviders.Abstractions": "5.0.0",
          "Microsoft.Extensions.FileProviders.Physical": "5.0.0"
        },
        "compile": {
          "lib/netstandard2.0/Microsoft.Extensions.Configuration.UserSecrets.dll": {}
        },
        "runtime": {
          "lib/netstandard2.0/Microsoft.Extensions.Configuration.UserSecrets.dll": {}
        },
        "build": {
          "build/netstandard2.0/Microsoft.Extensions.Configuration.UserSecrets.props": {},
          "build/netstandard2.0/Microsoft.Extensions.Configuration.UserSecrets.targets": {}
        }
      },
```

#### 6.0.0

```
      "Microsoft.Extensions.Configuration.UserSecrets/6.0.0": {
        "type": "package",
        "dependencies": {
          "Microsoft.Extensions.Configuration.Abstractions": "6.0.0",
          "Microsoft.Extensions.Configuration.Json": "6.0.0",
          "Microsoft.Extensions.FileProviders.Abstractions": "6.0.0",
          "Microsoft.Extensions.FileProviders.Physical": "6.0.0"
        },
        "compile": {
          "lib/netstandard2.0/Microsoft.Extensions.Configuration.UserSecrets.dll": {}
        },
        "runtime": {
          "lib/netstandard2.0/Microsoft.Extensions.Configuration.UserSecrets.dll": {}
        },
        "build": {
          "build/netstandard2.0/_._": {}
        }
      },
```

Just inspecting the package doesn't tell me what the issue could be, the `build/netstandard2.0/*.targets` file exists. But for some reason NuGet isn't respecting it.

@ericstj @safern - can you spot why our build logic isn't getting used for this package? Do you happen to know of a good way to debug why NuGet isn't picking it up?

**Comment by eerhardt** [comment]
I was able to figure out why the build assets aren't being included:

![image](https://user-images.githubusercontent.com/8291187/148242734-4d0d0ad5-5a39-476b-bb5a-a17cdd9bc204.png)

The reason is that the project only references `Microsoft.Extensions.Hosting` package.

https://github.com/devhl-labs/Hosting-6.0-Issue-Repro/blob/daabb5c610c640a76fae9d2bc0391fcc620424db/WorkerService1/WorkerService1/WorkerService1.csproj#L9

And between `5.0.0` and `6.0.0` we added `exclude="Build,Analyzers"` to all the dependencies in Microsoft.Extensions.Hosting. Thus when just referencing Microsoft.Extensions.Hosting, the UserSecrets build logic isn't getting imported. Thus, no UserSecretsIdAttribute is being written to the assembly.

This has actually broken UserSecrets in the Worker template as well. We should get this fixed in 6.0.x somehow. Here are the options I can think of:

1. Remove the `exclude="Build` on the UserSecrets dependency in Microsoft.Extensions.Hosting.
2. There is actually [logic in the SDK to inject the UserSecretsId attribute](https://github.com/dotnet/sdk/blob/22c4860dcb2cf6b123dd641cc4a87a50380759d5/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets#L443-L476). But that logic just looks for a `Microsoft.Extensions.Configuration.UserSecrets` PackageReference. We could update this to also look for a PackageReference to `Microsoft.Extensions.Hosting`.
3. We could update the template to also include a PackageReference to `Microsoft.Extensions.Configuration.UserSecrets`.
    - Note this option wouldn't fix existing projects

My opinion is we take the (1) approach above.

**Comment by eerhardt** [comment]
Related: https://github.com/dotnet/runtime/issues/36495#issuecomment-714596416

**Comment by eerhardt** [comment]
One thing that is unfortunate about fixing this issue with `​buildTransitive`​ is that in order for the Worker template to be fixed, 3 changes need to occur:

1. Fix the UserSecrets package as above.
2. Update `Microsoft.Extensions.Hosting` to depend on the new UserSecrets package. So we need to service the Hosting package to be `6.0.1`.
3. Update the template to reference the `6.0.1` package instead of `6.0.0`.

I think going forward the `buildTransitive` approach is the right thing to do.

**Comment by safern** [comment]
Fixed on main with: #63415 and on release/6.0 with: #63423
